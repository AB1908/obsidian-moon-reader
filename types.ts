export class Annotation {
    //todo: check types
    indexCount: number;
    bookName: string;
    bookPath: string;
    lowerCasePath: string;
    sectionNumber: number;
    placeholder1: string;
    location: string;
    characterCount: number;
    signedColor: number;
    unixTimestamp: string;
    bookmarkText: string;
    noteText: string;
    highlightText: string;
    annotType1: number;
    annotType2: number;
    annotType3: number;
}