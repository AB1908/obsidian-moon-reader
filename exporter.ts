import {Annotation} from 'types';
import integerToRGBA from './util';

export function generateOutput(
    listOfAnnotations: Annotation[]
): string {
    const sample = listOfAnnotations[0];
    console.log(sample);
    console.log(listOfAnnotations[1]);
    const frontmatter = `---\npath: ${sample.bookPath}\ntitle: ${sample.bookName}\nauthor: \n---\n`;
    var output = frontmatter;
    for (let annotation of listOfAnnotations) {
        var annotationAsString: string;
        if (annotation.highlightText) {
            annotationAsString = `\n:::\n\n`;
            annotationAsString += `color:: ${integerToRGBA(annotation.signedColor)}\n\n> ${annotation.highlightText}`;
            if (annotation.noteText) {
                annotationAsString += `\n\n${annotation.noteText}\n\n`;
                annotationAsString += ":::\n";
            } else {
                annotationAsString += `\n\n:::\n`;
            }
        }
        if (annotationAsString){
            output += annotationAsString;
        }
    }

    return output;
}