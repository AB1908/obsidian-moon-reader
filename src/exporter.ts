import { Annotation } from 'src/types';
import integerToRGBA from './util';
import { TFile } from 'obsidian';

export function generateOutput(listOfAnnotations: Annotation[], mrexptTFile: TFile, colorFilter: number): string {
    const sample = listOfAnnotations[0];
    //TODO: extract into template
    // TODO: last exported ID is likely broken
    const frontmatter = `---
path: "${mrexptTFile.path}"
title: "${sample.bookName}"
author: 
lastExportedTimestamp: ${mrexptTFile.stat.mtime}
lastExportedID: ${listOfAnnotations.last().indexCount}
---

`;
    let output = frontmatter;

    for (const annotation of listOfAnnotations.filter(t=>t.signedColor == colorFilter)) {
        let annotationAsString: string;
        if (annotation.highlightText) {
            annotationAsString = `${template(integerToRGBA(annotation.signedColor), annotation.highlightText, annotation.noteText)}\n`;
        }
        if (annotationAsString) {
            output += annotationAsString;
        }
    }

    return output;
}

function template(type: any, highlight: string, note: string) {
    if (highlight.includes("\n")) {
        highlight = highlight.replaceAll("\n", "\n> ");
    }
    return `> [!${type}]
> ${highlight}
> ***
> ${note}
`;
}
