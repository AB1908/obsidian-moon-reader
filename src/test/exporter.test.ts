import {expect, test} from "@jest/globals";
import {generateOutput} from "../exporter";
import {
	colorFilter,
	listOfAnnotations,
	mrexptTFile
} from "./json/howtotakenotes-success/exporter.generateOutput.input-2022-11-19 14-04-30.json";
import {Annotation} from "../types";
import {TFile} from "obsidian";
import {output as parsedMarkdown} from "./json/howtotakenotes-success/exporter.generateOutput.output-2022-11-19 14-04-30.json"

test("exporter.generateOutput", () => {
	const list = Array.from(listOfAnnotations);
	expect(generateOutput(list as Annotation[], mrexptTFile as unknown as TFile, colorFilter)).toBe(parsedMarkdown)
})

