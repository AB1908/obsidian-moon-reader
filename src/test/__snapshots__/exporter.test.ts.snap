// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exporter.generateOutput 1`] = `
"---
path: "Book Exports/Sönke Ahrens - How to Take Smart Notes_ One Simple Technique to Boost Writing,  Learning and Thinking-Sönke Ahrens (2022).mrexpt"
title: "How to Take Smart Notes. One Simple Technique to Boost Writing,  Learning and Thinking"
author: 
lastExportedTimestamp: 1665321164166
lastExportedID: 12623
---

> [!555555FF]
> INTRODUCTION
> ***
> 

> [!555555FF]
> 1 Everything You Need to Know
> ***
> ##

> [!555555FF]
> 2 Everything You Need to Do
> 
> ***
> ##

> [!555555FF]
> 3 Everything You Need to Have
> ***
> ##

> [!555555FF]
> 4 A Few Things to Keep in Mind
> ***
> ##

> [!555555FF]
> 5 Writing Is the Only Thing That Matters
> ***
> ##

> [!555555FF]
> 6 Simplicity Is Paramount
> ***
> ##

> [!555555FF]
> 7 Nobody Ever Starts From Scratch
> ***
> ##

> [!555555FF]
> 8 Let the Work Carry You Forward
> ***
> ##

> [!555555FF]
> 9 Separate and Interlocking Tasks
> ***
> ##

> [!555555FF]
> 10 Read for Understanding
> ***
> ##

> [!555555FF]
> 11 Take Smart Notes
> ***
> ##

> [!555555FF]
> 12 Develop Ideas
> ***
> ##

> [!555555FF]
> 13 Share Your Insight
> ***
> ##

> [!555555FF]
> 14 Make It a Habit
> ***
> ##

> [!555555FF]
> 1.1 Good Solutions are Simple – and Unexpected
> ***
> ###

> [!555555FF]
> 1.2 The Slip-box
> ***
> ###

> [!555555FF]
> 1.3 The slip-box manual
> ***
> ###

"
`;
